aten/src/ATen/native/cuda/Loss.cu

```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+    scalar_t* output,
+    scalar_t* total_weight,
+    scalar_t* input,
+    int64_t* target,
+    scalar_t* weights,
+    bool size_average,
+    int64_t nframe,
+    int64_t ndim,
+    int64_t n_classes,
+    int64_t ignore_index) {
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  __shared__ accscalar_t sh_inputs[N_THREADS], acc_weight[N_THREADS];
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  int64_t i, t;
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  scalar_t cur_weight;
+
+  sh_inputs[threadIdx.x] = static_cast<accscalar_t>(0);
+  acc_weight[threadIdx.x] = static_cast<accscalar_t>(0);
+  for (i = threadIdx.x; i < nframe; i += N_THREADS) {
+    t = target[i];
+    if (t != ignore_index) {
+      CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
+      cur_weight = weights != nullptr ? weights[t] : static_cast<scalar_t>(1);
+      sh_inputs[threadIdx.x] -= input[i * ndim + t] * cur_weight;
+      acc_weight[threadIdx.x] += cur_weight;
+    }
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    *output = *total_weight = static_cast<scalar_t>(0);
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (i = 0; i < N_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
+    *output = static_cast<scalar_t>(output_acc);
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (*total_weight != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
+    }
+  }
+}
+
+} // namespace
+
+void nll_loss_forward_out_cuda_template(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const c10::optional<Tensor>& weight_opt,
+    int64_t reduction,
+    int64_t ignore_index) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> weight_maybe_owned =
+      at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+
+  TensorArg output_arg{output, "output", 1};
+  TensorArg total_weight_arg{total_weight, "total_weight", 2};
+  TensorArg input_arg{input, "input", 3};
+  TensorArg target_arg{target, "target", 4};
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda",
+        {output_arg, total_weight_arg, input_arg, target_arg, weight_arg});
+  } else {
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda",
+        {output_arg, total_weight_arg, input_arg, target_arg});
+  }
+
+  const int64_t n_classes = input.size(-1);
+  const int64_t n_dims = input.dim();
+  TORCH_CHECK(n_dims > 0 && n_dims <= 2, "input tensor should be 1D or 2D");
+  TORCH_CHECK(
+      target.dim() == 1,
```

Comment:
What if target.dim() is 0? Previous code errored on this too

Reply:
I think the original implementation did not raise this specific error when `target.dim() == 0`:

https://github.com/pytorch/pytorch/blob/d5988c5eca0221e9ef58918e4f0b504940cb926a/aten/src/THCUNN/generic/ClassNLLCriterion.cu#L14-L16

I updated the check to `TORCH_CHECK(target.dim() <= 1)` to not change the original behavior.

Reply:
Can you please beef up testing for supported/unsupported dims?